src/query/ts/m3db/encoded_step_iterator.go

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/models"
+)
+
+type peekValue struct {
+	started  bool
+	finished bool
+	point    ts.Datapoint
+}
+
+type encodedStepIter struct {
+	mu           sync.RWMutex
+	lastBlock    bool
+	exhausted    bool
+	started      bool
+	currentTime  time.Time
+	err          error
+	bounds       models.Bounds
+	meta         block.Metadata
+	seriesMeta   []block.SeriesMeta
+	seriesIters  []encoding.SeriesIterator
+	seriesPeek   []peekValue
+	consolidator consolidator
+}
+
+func (b *encodedBlock) stepIter() block.StepIter {
+	cs := b.consolidation
+	consolidator := buildConsolidator(
+		time.Minute,
+		cs.bounds.StepSize,
+		cs.currentTime,
+		len(b.seriesBlockIterators),
+		cs.consolidationFn,
+	)
+	return &encodedStepIter{
+		lastBlock:    b.lastBlock,
+		currentTime:  cs.currentTime,
+		bounds:       cs.bounds,
+		meta:         b.meta,
+		seriesMeta:   b.seriesMetas,
+		seriesIters:  b.seriesBlockIterators,
+		consolidator: consolidator,
+	}
+}
+
+type encodedStep struct {
+	time   time.Time
+	values []float64
+}
+
+func (s *encodedStep) Time() time.Time   { return s.time }
+func (s *encodedStep) Values() []float64 { return s.values }
+
+func (it *encodedStepIter) Current() (block.Step, error) {
+	it.mu.RLock()
+	currentTime := it.currentTime
+	if !it.started || currentTime.After(it.bounds.End()) {
+		it.mu.RUnlock()
+		panic("out of bounds")
```

Comment:
return error ?

Reply:
I think iterators typically panic if you call `Current` while out of bounds, but can do either way

Reply:
Current() here already returns an error in the second arg, why do we have to force panic when we can avoid that

Reply:
I can change this but it doesn't really matter so much; will be intending to refactor how these works to mirror SeriesIterator, where neither `Next` or `Current` can error, but there's an `Err` method to check if any errors have occured